app/controllers/players_controller.rb

```diff
@@ -34,6 +34,17 @@ def update
 
   def destroy
     player = Player.find(params[:id])
+    history = PlayerHistory.from_player(player)
+
+    # Delete each match/game in this player's history
+    history.each do |match|
+      StatService.new({
+        match: match,
+        winner: match.winner,
+        loster: match.loser
+      }).update_stats_for_deleted_match
+      match.destroy
```

Comment:
We could do the same with `after_create` and `before_update`. We can also use [`ActiveModel::Dirty`](http://api.rubyonrails.org/classes/ActiveModel/Dirty.html) (which should already be included in the base model) to detect that the winner has changed in the `before_update` callback, so we won't even need to manually track the previous winner.